File: IAIso-v5.0/core/iaiso-php/tests/Unit/Audit/AuditTest.php
<?php

declare(strict_types=1);

namespace IAIso\Tests\Unit\Audit;

use IAIso\Audit\Event;
use IAIso\Audit\FanoutSink;
use IAIso\Audit\JsonlFileSink;
use IAIso\Audit\MemorySink;
use IAIso\Audit\NullSink;
use PHPUnit\Framework\TestCase;

final class AuditTest extends TestCase
{
    public function testEventEmitsFieldsInSpecOrder(): void
    {
        $e = new Event('exec-1', 'engine.init', 0.0, ['pressure' => 0.0]);
        $json = $e->toJson();
        // schema_version, execution_id, kind, timestamp, data
        self::assertStringStartsWith('{"schema_version":', $json);
        self::assertStringContainsString('"execution_id":"exec-1"', $json);
        $svPos = strpos($json, '"schema_version"');
        $eiPos = strpos($json, '"execution_id"');
        $knPos = strpos($json, '"kind"');
        $tsPos = strpos($json, '"timestamp"');
        $dtPos = strpos($json, '"data"');
        self::assertLessThan($eiPos, $svPos);
        self::assertLessThan($knPos, $eiPos);
        self::assertLessThan($tsPos, $knPos);
        self::assertLessThan($dtPos, $tsPos);
    }

    public function testIntegerFloatsSerializeAsIntegers(): void
    {
        $e = new Event('e', 'k', 0.0, ['n' => 0.0]);
        $json = $e->toJson();
        // 0.0 should appear as 0, not 0.0
        self::assertStringContainsString('"timestamp":0', $json);
        self::assertStringNotContainsString('"timestamp":0.0', $json);
        self::assertStringContainsString('"n":0', $json);
        self::assertStringNotContainsString('"n":0.0', $json);
    }

    public function testDataKeysSortedAlphabetically(): void
    {
        $e = new Event('e', 'k', 0.0, ['z' => 1, 'a' => 2, 'm' => 3]);
        $json = $e->toJson();
        $aPos = strpos($json, '"a"');
        $mPos = strpos($json, '"m"');
        $zPos = strpos($json, '"z"');
        self::assertLessThan($mPos, $aPos);
        self::assertLessThan($zPos, $mPos);
    }

    public function testNullDataValuesEmit(): void
    {
        $e = new Event('e', 'k', 0.0, ['tag' => null]);
        $json = $e->toJson();
        self::assertStringContainsString('"tag":null', $json);
    }

    public function testMemorySinkStoresEvents(): void
    {
        $sink = new MemorySink();
        $sink->emit(new Event('e', 'a', 0.0, []));
        $sink->emit(new Event('e', 'b', 0.0, []));
        self::assertCount(2, $sink->events());
        self::assertSame('a', $sink->events()[0]->kind);
    }

    public function testFanoutSinkBroadcasts(): void
    {
        $a = new MemorySink();
        $b = new MemorySink();
        $f = new FanoutSink($a, $b);
        $f->emit(new Event('e', 'k', 0.0, []));
        self::assertCount(1, $a->events());
        self::assertCount(1, $b->events());
    }

    public function testJsonlFileSinkAppends(): void
    {
        $path = tempnam(sys_get_temp_dir(), 'iaiso-audit-');
        $sink = new JsonlFileSink($path);
        $sink->emit(new Event('e', 'a', 0.0, []));
        $sink->emit(new Event('e', 'b', 0.0, []));
        $contents = file_get_contents($path);
        unlink($path);
        $lines = array_filter(explode("\n", $contents));
        self::assertCount(2, $lines);
        foreach ($lines as $line) {
            $obj = json_decode($line, true);
            self::assertIsArray($obj);
            self::assertArrayHasKey('schema_version', $obj);
        }
    }

    public function testNullSinkSwallowsEvents(): void
    {
        // Just confirms it doesn't blow up
        NullSink::instance()->emit(new Event('e', 'k', 0.0, []));
        self::assertTrue(true);
    }
}

File: IAIso-v5.0/core/iaiso-ruby/test/audit_test.rb
# frozen_string_literal: true

require_relative "test_helper"
require "iaiso/audit"
require "json"
require "tempfile"

class AuditTest < Minitest::Test
  def test_event_emits_fields_in_spec_order
    e = IAIso::Audit::Event.new(execution_id: "exec-1", kind: "engine.init", timestamp: 0.0, data: { "pressure" => 0.0 })
    json = e.to_json
    assert json.start_with?('{"schema_version":')
    assert_includes json, '"execution_id":"exec-1"'
    sv = json.index('"schema_version"')
    ei = json.index('"execution_id"')
    kn = json.index('"kind"')
    ts = json.index('"timestamp"')
    dt = json.index('"data"')
    assert sv < ei && ei < kn && kn < ts && ts < dt, "spec field order broken"
  end

  def test_integer_floats_serialize_as_integers
    e = IAIso::Audit::Event.new(execution_id: "e", kind: "k", timestamp: 0.0, data: { "n" => 0.0 })
    json = e.to_json
    assert_includes json, '"timestamp":0'
    refute_includes json, '"timestamp":0.0'
    assert_includes json, '"n":0'
    refute_includes json, '"n":0.0'
  end

  def test_data_keys_sorted_alphabetically
    e = IAIso::Audit::Event.new(execution_id: "e", kind: "k", timestamp: 0.0, data: { "z" => 1, "a" => 2, "m" => 3 })
    json = e.to_json
    a = json.index('"a"')
    m = json.index('"m"')
    z = json.index('"z"')
    assert a < m && m < z
  end

  def test_null_data_values_emit
    e = IAIso::Audit::Event.new(execution_id: "e", kind: "k", timestamp: 0.0, data: { "tag" => nil })
    assert_includes e.to_json, '"tag":null'
  end

  def test_memory_sink_stores_events
    s = IAIso::Audit::MemorySink.new
    s.emit(IAIso::Audit::Event.new(execution_id: "e", kind: "a", timestamp: 0.0))
    s.emit(IAIso::Audit::Event.new(execution_id: "e", kind: "b", timestamp: 0.0))
    assert_equal 2, s.events.size
    assert_equal "a", s.events[0].kind
  end

  def test_fanout_broadcasts
    a = IAIso::Audit::MemorySink.new
    b = IAIso::Audit::MemorySink.new
    f = IAIso::Audit::FanoutSink.new(a, b)
    f.emit(IAIso::Audit::Event.new(execution_id: "e", kind: "k", timestamp: 0.0))
    assert_equal 1, a.events.size
    assert_equal 1, b.events.size
  end

  def test_jsonl_file_sink_appends
    Tempfile.create("iaiso-audit-") do |tmp|
      sink = IAIso::Audit::JSONLFileSink.new(tmp.path)
      sink.emit(IAIso::Audit::Event.new(execution_id: "e", kind: "a", timestamp: 0.0))
      sink.emit(IAIso::Audit::Event.new(execution_id: "e", kind: "b", timestamp: 0.0))
      lines = File.readlines(tmp.path).map(&:strip).reject(&:empty?)
      assert_equal 2, lines.size
      lines.each do |l|
        obj = JSON.parse(l)
        assert obj.key?("schema_version")
      end
    end
  end

  def test_null_sink_swallows
    IAIso::Audit::NullSink.instance.emit(IAIso::Audit::Event.new(execution_id: "e", kind: "k", timestamp: 0.0))
    pass
  end
end
